### PR TITLE
Allow bytestring 0.11

### LINKED ISSUE #36: Support for bytestring 0.11
AFAIK no code changes are required – only the upper bound needs to be bumped.

### PR DESCRIPTION
**Pull request by Minoru**
`cabal test --enable-tests --constraint 'bytestring == 0.11.0.0' --allow-newer` passes. Tests currently can't be built without `--allow-newer` because `text` requires `bytestring < 0.11`, but [that bound have been bumped](https://github.com/haskell/text/pull/296) and should be available in the next release of `text`.

Fixes #36.

### CODE REVIEW

**Comment by Bodigrim** [comment]
Could you please check that it works as intended for GHC 7.10? Usages of`PS` contructor should have failed IIUC.
https://github.com/haskell/zlib/blob/866e10f283c2afe894bda620ed9051aee3a7227e/Codec/Compression/Zlib/Internal.hs#L478-L480

**Comment by Minoru** [comment]
You're right, this doesn't even build with GHC 7.10. However, I can't see a way for me to fix it because I can't run the necessary tests: I created a Docker container with Ubuntu Xenial that ships GHC 7.10.3 and cabal-install 1.22.6. That Cabal doesn't support `--allow-newer`, and I can't update to cabal-install 3.0 because that GHC can't build the newer Cabal library.

I could patch it and make sure it *builds* with GHC 7.10, but without running the tests I'm not confident I wouldn't be breaking something (as I have no idea about zlib's internals). Do you have any advise, @Bodigrim?

**Comment by Minoru** [comment]
I got something going with ghcup. Perhaps I *will* be able to test this after all.

**Comment by Bodigrim** [comment]
@Minoru yes, you can use `ghcup` to install GHC 7.10 + Cabal 3.2, or install packages from https://launchpad.net/~hvr/+archive/ubuntu/ghc manually. Then switch to `text-1.2.4.1-rc` using `cabal.project`:
```
packages: .

constraints:
  bytestring >=0.11

source-repository-package
  type: git
  location: https://github.com/haskell/text
  tag: v1.2.4.1-rc1
```

Alternatively, use `stack` with `lts-6.35` and `text-1.2.4.1-rc` in `extra-deps`.

----

I guess one can resolve incompatibility similar to @phadej's work in `text`: https://github.com/haskell/text/commit/21281e7a0284ddcbfeca4b344be1bc88166d0df0#diff-442e8eb6c18a99aea06678813934a508de3c1e4f07bffcda88e9fe84f5dbbc18

**Comment by Minoru** [comment]
Okay, I got myself a "build matrix" with GHC 7.10.3/8.10.2 and bytestring 0.10.12.0/0.11.0.0. Three out of four combinations build and pass tests, while the fourth (GHC 7.10.3 and bytestring-0.11.0.0) obviously doesn't.

<details>
<summary>My testing setup</summary>


I wrote a Dockerfile build-and-test.dockerfile:

```dockerfile
FROM ubuntu:20.04

RUN apt-get update && apt-get install -y curl build-essential curl libffi-dev libffi7 libgmp-dev libgmp10 libncurses-dev libncurses5 libtinfo5 zlib1g-dev git
WORKDIR /root
RUN mkdir -p /root/.ghcup/bin
ENV PATH=/root/.ghcup/bin:$PATH
RUN curl -o ghcup https://downloads.haskell.org/~ghcup/x86_64-linux-ghcup
RUN chmod +x ghcup
RUN ./ghcup upgrade
RUN rm -f ghcup

RUN ghcup install cabal
RUN cabal update

ARG ghc
RUN ghcup install ghc $ghc
RUN ghcup set ghc $ghc

WORKDIR /zlib
COPY zlib.cabal .
COPY cabal.project .

ARG cabal_flags
ARG bytestring
RUN cabal $cabal_flags build --jobs --enable-tests --only-dependencies --constrain "bytestring == $bytestring" --allow-newer=bytestring,text

COPY . .

RUN cabal $cabal_flags build --jobs --enable-tests --constrain "bytestring == $bytestring" --allow-newer=bytestring,text
RUN cabal $cabal_flags test --jobs --enable-tests --constrain "bytestring == $bytestring" --allow-newer=bytestring,text
```

I also created a cabal.project file that points at the `text`'s `master` branch — which we need in order for `text` to get built against `bytestring`-0.11:

```
source-repository-package
    type: git
    location: https://github.com/haskell/text.git

packages: .
```

I build the four quadrants of the "test matrix" as follows:

```
$ docker build -t zlib-7.10.3-0.10.12.0 --build-arg ghc=7.10.3 --build-arg cabal_flags=--ghc-option=-optl-no-pie --build-arg bytestring=0.10.12.0 -f build-and-test.dockerfile .
$ docker build -t zlib-7.10.3-0.11.0.0 --build-arg ghc=7.10.3 --build-arg cabal_flags=--ghc-option=-optl-no-pie --build-arg bytestring=0.11.0.0 -f build-and-test.dockerfile .
$ docker build -t zlib-8.10.2-0.10.12.0 --build-arg ghc=8.10.2 --build-arg bytestring=0.10.12.0 -f build-and-test.dockerfile .
$ docker build -t zlib-8.10.2-0.11.0.0 --build-arg ghc=8.10.2 --build-arg bytestring=0.11.0.0 -f build-and-test.dockerfile .
```

The expectation is that all four should get built without an issue.

</details>

Out of 9 places where `PS` is used, only three are covered by tests. I calculated that by going through them and either changing `offset` to 0 in constructors, or using `0` instead of `offset` after matching a pattern.

So I guess the first order of business is to write some more tests that would fail whenever `offset` is ignored.

**Comment by Bodigrim** [comment]
I do not want to discourage you from writing tests, but it could be difficult or impossible to cover remaining use cases of `PS`, because `offset` used to be equal to 0 most of the time, unless `ByteString` has been sliced. (And in `bytestring-0.11` offset is **always** 0).

Note that I'm not a maintainer of `zlib`, but from my perspective if you fix all use cases of `PS` in a uniform manner (e. g., using `withBS` combinator linked above), it should be quite sufficient.

**Comment by Bodigrim** [comment]
@Minoru do you need any help with this? If you got stuck, I can give it a try myself.

**Comment by Minoru** [comment]
I wrote a single test and lay that aside. I still intend to finish it though. Let's put a bit of a deadline on it then: if I don't push anything new by Monday, consider this PR up for grabs.

**Comment by Minoru** [comment]
That single test is as far as I could go — I couldn't figure out any more ways to cover the other lines. I also ended up copying ByteStringCompat from `text` almost verbatim. The only difference is that `mkBS` takes the `offset`, which helps avoid `plusForeignPtr` calls in the "client" code, i.e. in Coded.Compression.Zlib.Internal.

This passes my makeshift "test matrix" described above, and already passed CI, so I think this is ready for review.

@Bodigrim, thanks for all the hints and support with this!

File: Codec/Compression/Zlib/ByteStringCompat.hs
Comment (Bodigrim) [inline Codec/Compression/Zlib/ByteStringCompat.hs:2]: @phadej are you all right with this attribution?

**Comment by Bodigrim** [comment]
Note that Travis CI builds still uses `bytestring-0.10` and cannot serve as a public evidence. One has to set `constraint: bytestring >= 0.11` explicitly in `cabal.project`. 

 I do not have any Travis credits left, so prepared a GitHub Actions build atop of your changes, providing necessary evidences. 
https://github.com/Bodigrim/zlib/runs/1582643985?check_suite_focus=true
https://github.com/Bodigrim/zlib/tree/github-actions-with-bytestring-0.11

CC @dcoutts

File: Codec/Compression/Zlib/ByteStringCompat.hs
Comment (phadej) [reply Codec/Compression/Zlib/ByteStringCompat.hs:2]: Yes. That module doesn't really contain anything "copyright-able", just cruft.

**Comment by sjakobi** [comment]
Ping @dcoutts, @hvr! :) It would be great to have this merged and released!

**Comment by kozross** [comment]
Is there something blocking this merge @dcoutts @hvr?

**Review by emilypi — APPROVED** [review]
LGTM, i'll see if i can get a hold of one of (HVR | Dcoutts). Worst case i think I, andrew or simon can tiebreak this if the maintainers are unreachable.

**Comment by emilypi** [comment]
Merged. Thanks @Minoru!

**Pull request merged**
